app/vmui/packages/vmui/src/pages/ExploreAlerts/ExploreRules.tsx

```diff
@@ -0,0 +1,244 @@
+    return () => {
+      const newParams = new URLSearchParams(searchParams);
+      newParams.delete("group_id");
+      newParams.delete("rule_id");
+      newParams.delete("alert_id");
+      setSearchParams(newParams);
+      setModalOpen(false);
+      navigate({
+        hash: `#${id}`,
+      });
+    };
+  };
+
+  const allTypes: Set<string> = new Set();
+  const allStates: Set<string> = new Set();
+  const filteredGroups: Group[] = [];
+
+  const {
+    groups,
+    isLoading,
+    error,
+  } = useFetchGroups({ blockFetch: modalOpen });
+
+  const pageLoaded = !isLoading && !error && groups?.length;
+  const savedScrollTop = localStorage.getItem("scrollTop");
+
+  useEffect(() => {
+    if (pageLoaded) return;
+    if (location.hash) {
+      const target = document.querySelector(location.hash);
+      if (target) {
+        let parent = target.closest("details");
+        while (parent) {
+          parent.open = true;
+          if (!parent?.parentElement) return;
+          parent = parent.parentElement.closest("details");
+        }
+        target.scrollIntoView();
+      }
+    } else {
+      if (savedScrollTop) {
+        window.scrollTo(0, parseInt(savedScrollTop));
+      }
+      const updateScrollPosition = () => {
+        localStorage.setItem("scrollTop", (window.scrollY || 0).toString());
+      };
+      window.addEventListener("scroll", updateScrollPosition);
+      return () => {
+        window.removeEventListener("scroll", updateScrollPosition);
+      };
+    }
+  }, [location, savedScrollTop, pageLoaded]);
+
+  groups.forEach((group) => {
+    const filteredRules: APIRule[] = [];
+    const statesPerGroup: Record<string, number> = {};
+    group.rules.forEach((rule) => {
+      const ruleType = rule.type.charAt(0).toUpperCase() + rule.type.slice(1);
+      allTypes.add(ruleType);
+      if (types?.length && !types.includes(ruleType)) return;
+
+      const state = getState(rule);
+      const stateName = state.charAt(0).toUpperCase() + state.slice(1);
+      allStates.add(stateName);
+      if (states?.length && !states.includes(stateName)) return;
+
+      if (
+        searchInput &&
+        !rule.name.toLowerCase().includes(searchInput.toLowerCase()) &&
+        !group.name.toLowerCase().includes(searchInput.toLowerCase()) &&
+        !group.file.toLowerCase().includes(searchInput.toLowerCase())
+      )
+        return;
+
+      filteredRules.push(rule);
+      if (state !== "no match" && state !== "unhealthy" && state !== "firing" && state !== "pending")
+        return;
+
+      const count = state === "firing" || state === "pending" ? rule?.alerts?.length : 1;
+      if (stateName in statesPerGroup) {
+        statesPerGroup[stateName] += count;
+      } else {
+        statesPerGroup[stateName] = count;
+      }
+    });
+    if (filteredRules.length) {
+      const g = Object.assign({}, group);
+      g.rules = filteredRules;
+      g.states = statesPerGroup;
+      filteredGroups.push(g);
+    }
+  });
```

Comment:
Maybe we can move this logic to some `utils` file, and here use function call, as example:
```suggestion
  const { filteredGroups, allTypes, allStates } = useMemo(
     () => filterGroups(groups || [], types, states, searchInput),
  [groups, types, states, searchInput]
);
```
As for me it will be better for readability.